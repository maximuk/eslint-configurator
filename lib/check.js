"use strict";

const fs = require('fs');
const path = require('path');
const {
  getUsedPlugins,
  isDeprecated,
  loadPluginRules,
  getESLintRules,
  getPlugins,
  createCLIEngine,
} = require('./utils.js');

const usedPlugins = getUsedPlugins();

// check configs itself
try {
  createCLIEngine();
  usedPlugins.forEach(plugin => {
    createCLIEngine(plugin);
  });
} catch (error) {
  console.log(error.message.red);
  process.exit(1);
}

function checkRules(rules, userRules) {
  const userRulesTemp = { ...userRules };

  const messages = {
    deprecated: [],
    missing: [],
  };

  Object.keys(rules).forEach((rule) => {
    if (isDeprecated(rules[rule])) {
      if (userRulesTemp[rule]) {
        const replacedBy = rules[rule].meta
          && rules[rule].meta.docs
          && rules[rule].meta.docs.replacedBy
          && rules[rule].meta.docs.replacedBy.join(', ');
        messages.deprecated.push({
          rule,
          replacedBy
        });
        delete userRulesTemp[rule];
      }
    } else if (userRulesTemp[rule]) {
      delete userRulesTemp[rule];
    } else {
      messages.missing.push(rule);
    }
  });

  messages.redundant = Object.keys(userRulesTemp);

  Object.keys(messages).forEach((section) => {
    if (messages[section].length === 0) {
      delete messages[section];
    }
  })

  return Object.keys(messages).length ? messages : null;
}

function printMessages(messages) {
  let hasMessages = false;
  Object.keys(messages)
    .filter(pluginMessages => messages[pluginMessages])
    .forEach((pluginMessages) => {
      hasMessages = true;
      console.log(`\n${pluginMessages}:`.red.bold);
      Object.keys(messages[pluginMessages]).forEach((section) => {
        console.log(`  ${section} rules:`.red);
        messages[pluginMessages][section].forEach((rule) => {
          let message = '    ';
          if (section === 'deprecated') {
            message += rule.rule.red;
            if (rule.replacedBy) {
              message += ' (use ';
              message += rule.replacedBy.green;
              message += ' instead)';
            }
          } else {
            message += rule.red;
          }
          console.log(message);
        });
      });
    });

  return hasMessages;
}

// check deprecated, redundant and missing rules
const messages = {};
messages['eslint rules'] = checkRules(getESLintRules(), loadPluginRules('base'));

const plugins = getPlugins();
usedPlugins.forEach(plugin => {
  messages[`eslint-plugin-${plugin}`] = checkRules(
    plugins[plugin].rules,
    loadPluginRules(plugin)
  );
});

if (printMessages(messages)) {
  process.exit(1);
}
