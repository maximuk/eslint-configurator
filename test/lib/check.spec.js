const { BASE_CONFIG_NAME, PLUGINS_LIST } = require('../../lib/const.js');

const mockConfig = class {
  constructor(name) {
    this.name = name;

    this.fakeConstructor(name);
  }
};

jest.mock('colors/safe', () => ({
  bold: { red: str => str },
  green: str => str,
  red: str => str,
}));
jest.mock('../../lib/utils.js', () => ({}), { virtual: true });
jest.mock('../../lib/config.js', () => mockConfig, { virtual: true });

describe('./lib/check.js', () => {
  let processExit;
  let consoleLog;

  const usedPlugins = [BASE_CONFIG_NAME, PLUGINS_LIST[0]];

  beforeEach(() => {
    processExit = process.exit;
    process.exit = jest.fn();

    consoleLog = console.log;
    console.log = jest.fn();
  });

  afterEach(() => {
    process.exit = processExit;
    console.log = consoleLog;

    jest.resetModules();
  });

  it('should check all available configs', () => {
    const utils = require('../../lib/utils.js');

    utils.createCLIEngine = jest.fn((name) => {
      if (name !== BASE_CONFIG_NAME)
        throw new Error('createCLIEngine error');
    });
    mockConfig.prototype.fakeConstructor = function (name) {
      if (usedPlugins.includes(name)) {
        this.userConfig = {
          rules: {
            correctRule: 0,
            deprecatedRule1: 1,
            deprecatedRule2: 1,
            deprecatedRule3: 1,
            redundantRule: 2,
          },
        };
        this.rules = {
          correctRule: {},
          deprecatedRule1: { deprecated: true },
          deprecatedRule2: { deprecated: true, replacedBy: 'replacedByRule' },
          deprecatedRule3: { deprecated: true, replacedBy: 'replacedByRule', rule: 'rule' },
          deprecatedRule4: { deprecated: true },
          missingRule: {},
        };
      }
    };

    require('../../lib/check.js')();

    expect(utils.createCLIEngine).toHaveBeenCalledTimes(2);
    expect(utils.createCLIEngine.mock.calls).toEqual(
      usedPlugins.map(v => [v]),
    );

    expect(process.exit).toHaveBeenCalledTimes(2);
    expect(process.exit).toHaveBeenCalledWith(1);

    expect(console.log.mock.calls).toMatchSnapshot();
  });

  it('should not log any messages for correct config', () => {
    const utils = require('../../lib/utils.js');

    utils.createCLIEngine = jest.fn();
    mockConfig.prototype.fakeConstructor = function (name) {
      if (usedPlugins.includes(name)) {
        this.userConfig = {
          rules: {
            correctRule1: 0,
            correctRule2: 1,
            correctRule3: 2,
          },
        };
        this.rules = {
          correctRule1: {},
          correctRule2: {},
          correctRule3: {},
          deprecatedRule: { deprecated: true },
        };
      }
    };

    require('../../lib/check.js')();

    expect(utils.createCLIEngine).toHaveBeenCalledTimes(2);
    expect(utils.createCLIEngine.mock.calls).toEqual(
      usedPlugins.map(v => [v]),
    );

    expect(process.exit).toHaveBeenCalledTimes(0);

    expect(console.log).toHaveBeenCalledTimes(0);
  });
});
