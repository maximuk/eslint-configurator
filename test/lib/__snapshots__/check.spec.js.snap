// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`./lib/check.js should check all available configs 1`] = `
Array [
  Array [
    "createCLIEngine error",
  ],
  Array [
    "
eslint rules:",
  ],
  Array [
    "  deprecated rules:",
  ],
  Array [
    "    %s",
    "deprecatedRule1",
  ],
  Array [
    "    %s (use %s instead)",
    "deprecatedRule2",
    "replacedByRule",
  ],
  Array [
    "    %s (use %s instead)",
    "deprecatedRule3",
    "replacedByRule",
  ],
  Array [
    "  missing rules:",
  ],
  Array [
    "    %s",
    "missingRule",
  ],
  Array [
    "  redundant rules:",
  ],
  Array [
    "    %s",
    "redundantRule",
  ],
  Array [
    "
eslint-plugin-import:",
  ],
  Array [
    "  deprecated rules:",
  ],
  Array [
    "    %s",
    "deprecatedRule1",
  ],
  Array [
    "    %s (use %s instead)",
    "deprecatedRule2",
    "replacedByRule",
  ],
  Array [
    "    %s (use %s instead)",
    "deprecatedRule3",
    "replacedByRule",
  ],
  Array [
    "  missing rules:",
  ],
  Array [
    "    %s",
    "missingRule",
  ],
  Array [
    "  redundant rules:",
  ],
  Array [
    "    %s",
    "redundantRule",
  ],
]
`;
